File: sut/dotnet-api/Services/ItemStore.cs
using System.Collections.Concurrent;
using QaPortfolio.DotNetApi.Models;

namespace QaPortfolio.DotNetApi.Services;

public class ItemStore
{
    private readonly ConcurrentDictionary<Guid, ItemDto> _items = new();

    public ItemDto Create(ItemCreateDto req)
    {
        var id = Guid.NewGuid();
        var item = new ItemDto(
            id,
            req.Name.Trim(),
            req.Description ?? string.Empty,
            req.Active ?? true,
            DateTime.UtcNow);
        _items[id] = item;
        return item;
    }

    public ItemDto? Get(Guid id) => _items.TryGetValue(id, out var item) ? item : null;

    public ItemDto? Update(Guid id, ItemUpdateDto req)
    {
        if (!_items.TryGetValue(id, out var existing)) return null;
        var name = req.Name != null ? req.Name.Trim() : existing.Name;
        var description = req.Description ?? existing.Description;
        var active = req.Active ?? existing.Active;
        var updated = existing with { Name = name, Description = description, Active = active };
        _items[id] = updated;
        return updated;
    }

    public bool Delete(Guid id) => _items.TryRemove(id, out _);

    public PagedItemsDto List(int page, int size, bool? active)
    {
        IEnumerable<ItemDto> all = _items.Values;
        if (active.HasValue)
            all = all.Where(i => i.Active == active.Value);
        var sorted = all.OrderByDescending(i => i.CreatedAt).ToList();
        var total = sorted.Count;
        var totalPages = total == 0 ? 0 : (int)Math.Ceiling(total / (double)size);
        var content = sorted.Skip(page * size).Take(size).ToList();
        return new PagedItemsDto(content, page, size, total, totalPages);
    }
}

File: sut/spring-api/src/main/java/com/qaportfolio/springapi/service/ItemStore.java
package com.qaportfolio.springapi.service;

import com.qaportfolio.springapi.model.*;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

@Service
public class ItemStore {
    private final Map<UUID, Item> items = new ConcurrentHashMap<>();

    public Item create(ItemCreateRequest req) {
        UUID id = UUID.randomUUID();
        Item item = new Item(
                id,
                req.name().trim(),
                req.description() != null ? req.description() : "",
                req.active() == null || req.active(),
                Instant.now()
        );
        items.put(id, item);
        return item;
    }

    public Optional<Item> findById(UUID id) {
        return Optional.ofNullable(items.get(id));
    }

    public Optional<Item> update(UUID id, ItemUpdateRequest req) {
        Item existing = items.get(id);
        if (existing == null) return Optional.empty();
        String name = req.name() != null ? req.name().trim() : existing.name();
        String description = req.description() != null ? req.description() : existing.description();
        boolean active = req.active() != null ? req.active() : existing.active();
        Item updated = new Item(id, name, description, active, existing.createdAt());
        items.put(id, updated);
        return Optional.of(updated);
    }

    public boolean delete(UUID id) {
        return items.remove(id) != null;
    }

    public PagedItems list(int page, int size, Boolean active) {
        List<Item> all = new ArrayList<>(items.values());
        if (active != null) {
            all = all.stream().filter(i -> i.active() == active).collect(Collectors.toList());
        }
        all.sort(Comparator.comparing(Item::createdAt).reversed());
        long total = all.size();
        int totalPages = total == 0 ? 0 : (int) Math.ceil((double) total / size);
        int from = page * size;
        List<Item> content = from >= all.size()
                ? List.of()
                : all.subList(from, Math.min(from + size, all.size()));
        return new PagedItems(content, page, size, total, totalPages);
    }
}
